File: CortexFrontend/src/utils/catalogSearchRanking.ts
/**
 * Catalog search ranking for Configuration visibility (mirror of CortexBackend.Services.CatalogSearchRanking).
 * Keep behavior aligned when updating either side.
 */

import type { SapReferenceCatalogEntry } from "../types/sapReferenceCatalog";
import type { SynitiKnowledgeCatalogEntry } from "../types/synitiKnowledgeCatalog";

const delimiterPattern = /[|;\n\r]+/g;

export function normalizeSearchText(input?: string | null): string {
  if (!input?.trim()) {
    return "";
  }
  return input
    .trim()
    .toLowerCase()
    .replace(/\s+/g, " ");
}

function splitPhrases(raw: string | null | undefined): string[] {
  if (!raw?.trim()) {
    return [];
  }
  const parts: string[] = [];
  for (const segment of raw.split(delimiterPattern)) {
    const n = normalizeSearchText(segment);
    if (n.length > 0) {
      parts.push(n);
    }
  }
  return parts;
}

function phraseListContainsSubstring(raw: string | null | undefined, q: string): boolean {
  if (!raw?.trim()) {
    return false;
  }
  for (const phrase of splitPhrases(raw)) {
    if (phrase !== q && phrase.includes(q)) {
      return true;
    }
  }
  return false;
}

function containsNormalized(text: string | null | undefined, q: string): boolean {
  if (!text?.trim()) {
    return false;
  }
  return normalizeSearchText(text).includes(q);
}

function guidanceOrRelatedContains(entry: SynitiKnowledgeCatalogEntry, q: string): boolean {
  if (
    containsNormalized(entry.businessMeaning, q) ||
    containsNormalized(entry.shortDefinition, q) ||
    containsNormalized(entry.technicalMeaning, q) ||
    containsNormalized(entry.relatedTerms, q)
  ) {
    return true;
  }
  for (const line of entry.suggestedReviewerChecks) {
    if (normalizeSearchText(line).includes(q)) {
      return true;
    }
  }
  for (const line of entry.missingContextQuestions) {
    if (normalizeSearchText(line).includes(q)) {
      return true;
    }
  }
  return false;
}

export function getSynitiSortKey(entry: SynitiKnowledgeCatalogEntry, normalizedQuery: string): number {
  if (normalizedQuery.length === 0) {
    return 0;
  }

  const term = normalizeSearchText(entry.term);
  const category = normalizeSearchText(entry.category);
  let best = Number.MAX_SAFE_INTEGER;

  if (term === normalizedQuery) {
    best = Math.min(best, 100);
  }
  if (term.startsWith(normalizedQuery)) {
    best = Math.min(best, 200);
  }
  if (category === normalizedQuery) {
    best = Math.min(best, 300);
  }
  for (const phrase of splitPhrases(entry.aliases)) {
    if (phrase === normalizedQuery) {
      best = Math.min(best, 300);
    }
  }
  for (const phrase of splitPhrases(entry.examplePhrases)) {
    if (phrase === normalizedQuery) {
      best = Math.min(best, 300);
    }
  }
  if (term !== normalizedQuery && !term.startsWith(normalizedQuery) && term.includes(normalizedQuery)) {
    best = Math.min(best, 400);
  }
  if (
    phraseListContainsSubstring(entry.aliases, normalizedQuery) ||
    phraseListContainsSubstring(entry.examplePhrases, normalizedQuery)
  ) {
    best = Math.min(best, 500);
  }
  if (
    containsNormalized(entry.aliases, normalizedQuery) ||
    containsNormalized(entry.examplePhrases, normalizedQuery)
  ) {
    best = Math.min(best, 500);
  }
  if (guidanceOrRelatedContains(entry, normalizedQuery)) {
    best = Math.min(best, 600);
  }

  return best === Number.MAX_SAFE_INTEGER ? 650 : best;
}

function contextExactMatch(entry: SapReferenceCatalogEntry, q: string): boolean {
  const m = normalizeSearchText(entry.module);
  const d = normalizeSearchText(entry.domain);
  const b = normalizeSearchText(entry.businessObject);
  return (m.length > 0 && m === q) || (d.length > 0 && d === q) || (b.length > 0 && b === q);
}

function contextSubstringMatch(entry: SapReferenceCatalogEntry, q: string): boolean {
  for (const p of [entry.module, entry.domain, entry.businessObject]) {
    const n = normalizeSearchText(p);
    if (n.length > 0 && n.includes(q)) {
      return true;
    }
  }
  return false;
}

function descriptionOrContextContains(entry: SapReferenceCatalogEntry, q: string): boolean {
  return (
    containsNormalized(entry.tableDescription, q) ||
    containsNormalized(entry.fieldDescription, q) ||
    contextSubstringMatch(entry, q)
  );
}

function sourceFieldsContain(entry: SapReferenceCatalogEntry, q: string): boolean {
  return containsNormalized(entry.sourceName, q) || containsNormalized(entry.sourceType, q);
}

export function getSapSortKey(entry: SapReferenceCatalogEntry, normalizedQuery: string): number {
  if (normalizedQuery.length === 0) {
    return 0;
  }

  const table = normalizeSearchText(entry.tableName);
  const field = entry.fieldName ? normalizeSearchText(entry.fieldName) : "";
  const isTable = entry.rowKind.toLowerCase() === "table";

  let best = Number.MAX_SAFE_INTEGER;

  if (isTable && table === normalizedQuery) {
    best = Math.min(best, 100);
  }
  if (!isTable && field === normalizedQuery) {
    best = Math.min(best, 110);
  }
  if (!isTable && table === normalizedQuery && field.length > 0 && field !== normalizedQuery) {
    best = Math.min(best, 120);
  }

  const startsTable = table.startsWith(normalizedQuery) && table !== normalizedQuery;
  const startsField = field.length > 0 && field.startsWith(normalizedQuery) && field !== normalizedQuery;
  if (startsTable || startsField) {
    const sub = isTable ? 0 : 1;
    best = Math.min(best, 200 + sub);
  }

  if (contextExactMatch(entry, normalizedQuery)) {
    best = Math.min(best, 300);
  }

  const nameContains =
    (table.includes(normalizedQuery) && table !== normalizedQuery && !table.startsWith(normalizedQuery)) ||
    (field.length > 0 &&
      field.includes(normalizedQuery) &&
      field !== normalizedQuery &&
      !field.startsWith(normalizedQuery));
  if (nameContains) {
    best = Math.min(best, 400);
  }

  if (descriptionOrContextContains(entry, normalizedQuery)) {
    best = Math.min(best, 500);
  }
  if (sourceFieldsContain(entry, normalizedQuery)) {
    best = Math.min(best, 600);
  }

  return best;
}

File: CortexBackend/Services/CatalogSearchRanking.cs
using System.Text.RegularExpressions;
using Cortex.API.DTO;

namespace Cortex.API.Services;

/// <summary>Deterministic search ranking for read-only catalog visibility (lower sort key = stronger match).</summary>
public static class CatalogSearchRanking
{
    private static readonly char[] Delimiters = ['|', ';', '\n', '\r'];

    public static string NormalizeSearchText(string? input)
    {
        if (string.IsNullOrWhiteSpace(input))
        {
            return string.Empty;
        }

        var s = input.Trim().ToLowerInvariant();
        return Regex.Replace(s, @"\s+", " ", RegexOptions.CultureInvariant);
    }

    /// <summary>When <paramref name="normalizedQuery"/> is empty, returns 0 (caller should sort by term only).</summary>
    public static int GetSynitiSortKey(SynitiKnowledgeCatalogEntryDto dto, string normalizedQuery)
    {
        if (normalizedQuery.Length == 0)
        {
            return 0;
        }

        var term = NormalizeSearchText(dto.Term);
        var category = NormalizeSearchText(dto.Category);
        var best = int.MaxValue;

        if (term == normalizedQuery)
        {
            best = Math.Min(best, 100);
        }

        if (term.StartsWith(normalizedQuery, StringComparison.Ordinal))
        {
            best = Math.Min(best, 200);
        }

        if (category == normalizedQuery)
        {
            best = Math.Min(best, 300);
        }

        foreach (var phrase in SplitPhrases(dto.Aliases))
        {
            if (phrase == normalizedQuery)
            {
                best = Math.Min(best, 300);
            }
        }

        foreach (var phrase in SplitPhrases(dto.ExamplePhrases))
        {
            if (phrase == normalizedQuery)
            {
                best = Math.Min(best, 300);
            }
        }

        if (term.Contains(normalizedQuery, StringComparison.Ordinal) && term != normalizedQuery &&
            !term.StartsWith(normalizedQuery, StringComparison.Ordinal))
        {
            best = Math.Min(best, 400);
        }

        if (PhraseListContainsSubstring(dto.Aliases, normalizedQuery) ||
            PhraseListContainsSubstring(dto.ExamplePhrases, normalizedQuery))
        {
            best = Math.Min(best, 500);
        }

        if (ContainsNormalized(dto.Aliases, normalizedQuery) ||
            ContainsNormalized(dto.ExamplePhrases, normalizedQuery))
        {
            best = Math.Min(best, 500);
        }

        if (GuidanceOrRelatedContains(dto, normalizedQuery))
        {
            best = Math.Min(best, 600);
        }

        return best == int.MaxValue ? 650 : best;
    }

    /// <summary>When <paramref name="normalizedQuery"/> is empty, returns 0 (caller uses default catalog order).</summary>
    public static int GetSapSortKey(SapReferenceCatalogEntryDto dto, string normalizedQuery)
    {
        if (normalizedQuery.Length == 0)
        {
            return 0;
        }

        var table = NormalizeSearchText(dto.TableName);
        var field = dto.FieldName != null ? NormalizeSearchText(dto.FieldName) : string.Empty;
        var isTable = dto.RowKind.Equals("Table", StringComparison.OrdinalIgnoreCase);

        var best = int.MaxValue;

        if (isTable && table == normalizedQuery)
        {
            best = Math.Min(best, 100);
        }

        if (!isTable && field == normalizedQuery)
        {
            best = Math.Min(best, 110);
        }

        if (!isTable && table == normalizedQuery && field.Length > 0 && field != normalizedQuery)
        {
            best = Math.Min(best, 120);
        }

        var startsTable = table.StartsWith(normalizedQuery, StringComparison.Ordinal) &&
                          table != normalizedQuery;
        var startsField = field.Length > 0 &&
                           field.StartsWith(normalizedQuery, StringComparison.Ordinal) &&
                           field != normalizedQuery;
        if (startsTable || startsField)
        {
            var sub = isTable ? 0 : 1;
            best = Math.Min(best, 200 + sub);
        }

        if (ContextExactMatch(dto, normalizedQuery))
        {
            best = Math.Min(best, 300);
        }

        var nameContains =
            (table.Contains(normalizedQuery, StringComparison.Ordinal) && table != normalizedQuery &&
             !table.StartsWith(normalizedQuery, StringComparison.Ordinal)) ||
            (field.Length > 0 && field.Contains(normalizedQuery, StringComparison.Ordinal) &&
             field != normalizedQuery && !field.StartsWith(normalizedQuery, StringComparison.Ordinal));
        if (nameContains)
        {
            best = Math.Min(best, 400);
        }

        if (DescriptionOrContextContains(dto, normalizedQuery))
        {
            best = Math.Min(best, 500);
        }

        if (SourceFieldsContain(dto, normalizedQuery))
        {
            best = Math.Min(best, 600);
        }

        return best;
    }

    public static bool SapEntryMatchesSearch(SapReferenceCatalogEntryDto dto, string normalizedQuery)
    {
        if (normalizedQuery.Length == 0)
        {
            return true;
        }

        return GetSapSortKey(dto, normalizedQuery) < int.MaxValue;
    }

    private static bool ContextExactMatch(SapReferenceCatalogEntryDto dto, string q)
    {
        var m = NormalizeSearchText(dto.Module);
        var d = NormalizeSearchText(dto.Domain);
        var b = NormalizeSearchText(dto.BusinessObject);
        return (m.Length > 0 && m == q) || (d.Length > 0 && d == q) || (b.Length > 0 && b == q);
    }

    private static bool DescriptionOrContextContains(SapReferenceCatalogEntryDto dto, string q)
    {
        return ContainsNormalized(dto.TableDescription, q) ||
               ContainsNormalized(dto.FieldDescription, q) ||
               ContextSubstringMatch(dto, q);
    }

    private static bool ContextSubstringMatch(SapReferenceCatalogEntryDto dto, string q)
    {
        var parts = new[] { dto.Module, dto.Domain, dto.BusinessObject };
        foreach (var p in parts)
        {
            var n = NormalizeSearchText(p);
            if (n.Length > 0 && n.Contains(q, StringComparison.Ordinal))
            {
                return true;
            }
        }

        return false;
    }

    private static bool SourceFieldsContain(SapReferenceCatalogEntryDto dto, string q)
    {
        return ContainsNormalized(dto.SourceName, q) || ContainsNormalized(dto.SourceType, q);
    }

    private static bool GuidanceOrRelatedContains(SynitiKnowledgeCatalogEntryDto dto, string q)
    {
        if (ContainsNormalized(dto.BusinessMeaning, q) ||
            ContainsNormalized(dto.ShortDefinition, q) ||
            ContainsNormalized(dto.TechnicalMeaning, q) ||
            ContainsNormalized(dto.RelatedTerms, q))
        {
            return true;
        }

        foreach (var line in dto.SuggestedReviewerChecks)
        {
            if (NormalizeSearchText(line).Contains(q, StringComparison.Ordinal))
            {
                return true;
            }
        }

        foreach (var line in dto.MissingContextQuestions)
        {
            if (NormalizeSearchText(line).Contains(q, StringComparison.Ordinal))
            {
                return true;
            }
        }

        return false;
    }

    private static bool PhraseListContainsSubstring(string? raw, string q)
    {
        if (string.IsNullOrWhiteSpace(raw))
        {
            return false;
        }

        foreach (var phrase in SplitPhrases(raw))
        {
            if (phrase != q && phrase.Contains(q, StringComparison.Ordinal))
            {
                return true;
            }
        }

        return false;
    }

    private static bool ContainsNormalized(string? text, string q)
    {
        if (string.IsNullOrWhiteSpace(text))
        {
            return false;
        }

        return NormalizeSearchText(text).Contains(q, StringComparison.Ordinal);
    }

    private static IEnumerable<string> SplitPhrases(string? raw)
    {
        if (string.IsNullOrWhiteSpace(raw))
        {
            yield break;
        }

        foreach (var segment in raw.Split(Delimiters, StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries))
        {
            var n = NormalizeSearchText(segment);
            if (n.Length > 0)
            {
                yield return n;
            }
        }
    }
}
